### PR TITLE
feat!: All parameter propagation (1/N)

### PR DESCRIPTION
**Pull request by asalzburger**
This PR is the first step towards all parameter propagation.

At current it does not change functionality, it does prepare the Stepper states for all propagation types:

free -> free
bound -> bound 
bound -> free
free -> bound

To do so, all Steppers have now variant types for the relevant covariances, jacobians.

### CODE REVIEW

**Comment by asalzburger** [comment]
UnitTests are still failing here - I will also make a timing comparison.

File: Core/include/Acts/Propagator/EigenStepper.hpp [hunk 1]
```diff
@@ -47,12 +49,32 @@ template <typename bfield_t,
...
-  using BoundState = std::tuple<BoundTrackParameters, Jacobian, double>;
+
+  /// Variant Jacobian:
+  /// free-bound, bound-bound, bound-free, free-free
+  using VariantJacobian = std::variant<FreeToBoundMatrix, BoundMatrix,
+                                       BoundToFreeMatrix, FreeMatrix>;
+
+  /// Variant Jacobian initial to global:
+  /// bound-free, free-free
+  using VariantJacobianToGlobal = std::variant<BoundToFreeMatrix, FreeMatrix>;
```
Comment (FabianKlimpel) [inline Core/include/Acts/Propagator/EigenStepper.hpp:61]: If you start in free parameters then you don't need this matrix. What's the benefit of overallocating memory compared to an `std::optional` that carries the same information?

Comment (asalzburger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:61]: It's true - I might change that to `std::optional`

**Comment by asalzburger** [comment]
Superseded by #744.

**Pull request closed without merging**